### PR TITLE
Save invoices on server during generation

### PR DESCRIPTION
**Pull request by Zales0123**
To keep invoices files in exactly the same state (in both template and data) as in the moment of the order being paid, we need to save them on the server, rather than creating them on the fly.

This PR also includes some refactoring and code improvements, to leave the code in a better shape than it was when I started to work with it 💃

### CODE REVIEW

File: src/Provider/InvoiceFileProvider.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        $this->invoicesDirectory = $invoicesDirectory;
+    }
+
+    public function provide(InvoiceInterface $invoice): InvoicePdf
+    {
+        $invoiceFileName = $this->invoiceFileNameGenerator->generateForPdf($invoice);
+        $invoiceFile = $this->filesystem->get($invoiceFileName);
+
+        $invoicePdf = new InvoicePdf($invoiceFileName, $invoiceFile->getContent());
+        $invoicePdf->setFullPath($this->invoicesDirectory . '/' . $invoiceFileName);
```
Comment (AdamKasp) [inline src/Provider/InvoiceFileProvider.php:48]: IMHO if we have more than one concatenation we should consider using`sprintf`

File: tests/Behat/Resources/suites/admin/managing_invoices.yml [hunk 2]
```diff
@@ -62,9 +63,57 @@ default:
...
+                - sylius.behat.context.setup.channel
+                - sylius.behat.context.setup.currency
+                - sylius.behat.context.setup.customer
+                - sylius.behat.context.setup.geographical
+                - sylius.behat.context.setup.order
+                - sylius.behat.context.setup.payment
+                - sylius.behat.context.setup.product
+                - sylius.behat.context.setup.product_taxon
+                - sylius.behat.context.setup.promotion
+                - sylius.behat.context.setup.admin_security
```
Comment (AdamKasp) [inline tests/Behat/Resources/suites/admin/managing_invoices.yml:80]: Do we really need all of these contexts?

File: src/Provider/InvoiceFileProvider.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    ) {
+        $this->invoiceFileNameGenerator = $invoiceFileNameGenerator;
+        $this->filesystem = $filesystem;
+        $this->invoicesDirectory = $invoicesDirectory;
+    }
+
+    public function provide(InvoiceInterface $invoice): InvoicePdf
+    {
+        $invoiceFileName = $this->invoiceFileNameGenerator->generateForPdf($invoice);
+        $invoiceFile = $this->filesystem->get($invoiceFileName);
```
Comment (GSadee) [inline src/Provider/InvoiceFileProvider.php:45]: What if there is no invoice saved in the filesystem? Shouldn't it be generated here?

File: UPGRADE.md [hunk 4]
```diff
@@ -1,3 +1,16 @@
+### UPGRADE FROM 0.16.1 TO 0.17.0
+
+Invoices are now saved on the server during their generation (by default, when the order is paid).
+
+1. `Sylius\InvoicingPlugin\Creator\InvoiceCreator` class has 2 more dependencies: `InvoicePdfFileGeneratorInterface $invoicePdfFileGenerator`
```
Comment (AdamKasp) [inline UPGRADE.md:5]: I am wondering if it will more clear if we show all constructor parameters instead of describing what have we changed.

**Review by AdamKasp — APPROVED** [review]

**Review by GSadee — APPROVED** [review]

File: src/Provider/InvoiceFileProvider.php (hunk 3 above)
Comment (GSadee) [reply src/Provider/InvoiceFileProvider.php:45]: And what is more, the current solution is not backwards compatible, as there will be no generated pdfs in the filesystem, so no one will be able to download the previously created invoice after this change.

Comment (Zales0123) [reply src/Provider/InvoiceFileProvider.php:45]: Agreed 👍 It should probably be generated here, to keep it both simple enough and backward compatible

File: features/managing_invoices/saving_invoices_on_server_during_generation.feature [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        And the store allows paying with "Cash on Delivery"
+        And channel "United States" has shop billing data set as "Ragnarok", "1100110011", "Pacific Coast Hwy", "90806" "Los Angeles", "United States"
+
+    @application
+    Scenario: Having invoice saved on the server after the order is placed
+        Given there is a customer "lucy@teamlucifer.com" that placed an order "#00000666"
+        When the customer bought 2 "Angel T-Shirt" products
+        And the customer "Lucifer Morningstar" addressed it to "Seaside Fwy", "90802" "Los Angeles" in the "United States"
+        And for the billing address of "Mazikeen Lilim" in the "Pacific Coast Hwy", "90806" "Los Angeles", "United States"
+        And the customer chose "UPS" shipping method with "Cash on Delivery" payment
```
[the remark this answers is not in the document]
Comment (Zales0123) [reply features/managing_invoices/saving_invoices_on_server_during_generation.feature:22]: Answer to both this comment and the one above 💃 cc @AdamKasp 

When putting these steps into the background/writing them with the `Given` keyword, we're missing _an action_ in this scenario, which is checked by the _assertion_ in the last step. The reason for that is how our steps for completing the checkout are named. And yes, they definitely seem like setting up the application 🚀 But renaming them a little bit too big work for a small benefit, at least in my opinion. 

So TL;DR - I wanted to emphasize the action that triggers the invoice generation and that's why all of these steps are here and marked as `When` rather than `Given` and/or in the background 🖖

File: src/Provider/InvoiceFileProvider.php (hunk 1 above)
Comment (Zales0123) [reply src/Provider/InvoiceFileProvider.php:48]: In general, I agree, but in this specific case it would look like:

```php
sprintf('%s/%s', $this->invoicesDirectory, $invoiceFileName)
```

Which is far less readable than the proposed concatenation in my opinion :D

File: tests/Behat/Resources/suites/admin/managing_invoices.yml (hunk 2 above)
Comment (Zales0123) [reply tests/Behat/Resources/suites/admin/managing_invoices.yml:80]: Probably I could remove some of them 👌

File: UPGRADE.md (hunk 4 above)
Comment (Zales0123) [reply UPGRADE.md:5]: 🤷‍♂️ both works for me 😄

**Review by lchrusciel — APPROVED** [review]

File: spec/Model/InvoicePdfSpec.php [hunk 6]
```diff
@@ -31,4 +31,10 @@ public function it_has_content(): void
     {
         $this->content()->shouldReturn('pdf content');
     }
+
+    function it_has_full_path(): void
```
Comment (GSadee) [inline spec/Model/InvoicePdfSpec.php:35]:
```suggestion
    function it_has_a_full_path(): void
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Mateusz! :tada:
